andi/andi.py

```diff
@@ -1,62 +1,327 @@
 
-    ``arguments_or_func`` should be either a callable, or
-    a result of andi.inspect call for a callable.
 
-    ``can_provide`` can be either a function which receives a type and
-    returns True if argument of such type can be provided, or a container
-    (e.g. a set) with supported types.
+Step = Tuple[Callable, KwargsSpec]
+
+
+class Plan(List[Step]):
     """
-    if callable(arguments_or_func):
-        arguments = inspect(arguments_or_func)
-    else:
-        arguments = arguments_or_func
+    The resultant plan of executing the ``plan`` function.
+    It contains a sequence of steps that should be executed in order because
+    they are dependant among them.
+
+    The plan is itself a list of tuples of type
+    ``Tuple[Callable, Dict[str, Callable]]``
+    corresponding to
+    (callable_to_invoke, (param_name -> callable_to_build_the_param)).
+    """
+
+    @property
+    def dependencies(self) -> List[Step]:
+        """
+        The plan required to build the dependencies for the
+        ``plan`` input function/class only.
+        Useful when it is known that not all dependencies for input
+        function/class could be resolved. In such a case, it is convenient to
+        execute the plan for the dependencies and then
+        have a custom execution step for the
+        input function/class where those non-resolvable dependencies are
+        provided by other means.
+
+        ``plan.dependencies`` is the ``plan`` without the last item ``plan[:-1]``
+        """
+        return self[:-1]
+
+    def final_kwargs(self, instances: Mapping[Callable, Any]) -> Dict[str, Any]:
+        """
+        Build the kwargs dict required to invoke the class/function
+        for which the plan was done for.
+        Equivalent to ``plan[-1][1].kwargs(instances)``
+
+        :param instances: A dict with the already prebuilt dependencies keyed
+                          by its builder
+        :return: a kwargs dict ready to be passed to a callable
+        """
+        return self[-1][1].kwargs(instances)
+
+
+def plan(class_or_func: Callable, *,
+         is_injectable: ContainerOrCallableType,
+         externally_provided: Optional[ContainerOrCallableType] = None,
+         full_final_arguments=False) -> Plan:
+    """ Plan the sequence of instantiation steps required to fulfill the
+    the arguments of the given function or the arguments of its
+    constructor if a class is given instead. In other words, this function
+    makes dependency injection easy. Type annotations are used
+    to determine with instance must be built to satisfy the dependency.
+
+    The plan is a sequence steps.
+    Each step in the plan is a tuple with:
+
+    * A callable with the
+      class/function that must be built/invoked in this step
+    * A dictionary with all the kwargs required for the
+      key build/invocation process. This dictionary has the argument names as keys
+      and classes/functions required to build them as values.
```

Comment:
I think it makes sense to mention KwargsSpec here, not just say that it is a dictionary.

Reply:
I missed to update all this docstring. Fixed here: https://github.com/scrapinghub/andi/pull/8/commits/080ae2734c4e79e241d80c5a72745fd65e4fdf3c

Note that I have also renamed `full_final_arguments` to `full_final_kwargs`. I think the meaning will be more clear now.